File: libs/ml-kernel/src/models/interop.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::models::run::{ArtifactReference, MetricValue};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ExternalTrackingSource {
    #[serde(default)]
    pub system: String,
    #[serde(default)]
    pub project: String,
    #[serde(default)]
    pub experiment_name: String,
    #[serde(default)]
    pub run_id: String,
    #[serde(default)]
    pub run_name: String,
    #[serde(default)]
    pub run_uri: String,
    #[serde(default)]
    pub artifact_uri: String,
    #[serde(default)]
    pub model_uri: String,
    #[serde(default)]
    pub registered_model_name: String,
    #[serde(default)]
    pub registered_model_version: String,
    #[serde(default)]
    pub framework: String,
    #[serde(default)]
    pub flavor: String,
    #[serde(default)]
    pub stage: String,
    #[serde(default)]
    pub tags: Value,
    #[serde(default)]
    pub params: Value,
    #[serde(default)]
    pub metrics: Vec<MetricValue>,
    #[serde(default)]
    pub artifacts: Vec<ArtifactReference>,
    #[serde(default)]
    pub metadata: Value,
}

impl ExternalTrackingSource {
    pub fn has_signal(&self) -> bool {
        !self.system.trim().is_empty()
            || !self.project.trim().is_empty()
            || !self.experiment_name.trim().is_empty()
            || !self.run_id.trim().is_empty()
            || !self.run_name.trim().is_empty()
            || !self.run_uri.trim().is_empty()
            || !self.artifact_uri.trim().is_empty()
            || !self.model_uri.trim().is_empty()
            || !self.registered_model_name.trim().is_empty()
            || !self.registered_model_version.trim().is_empty()
            || !self.framework.trim().is_empty()
            || !self.flavor.trim().is_empty()
            || !self.stage.trim().is_empty()
            || !self.metrics.is_empty()
            || !self.artifacts.is_empty()
            || !self.params.is_null()
            || !self.tags.is_null()
            || !self.metadata.is_null()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ModelAdapterDescriptor {
    #[serde(default)]
    pub kind: String,
    #[serde(default)]
    pub framework: String,
    #[serde(default)]
    pub flavor: String,
    #[serde(default)]
    pub runtime: String,
    #[serde(default)]
    pub loader: String,
    #[serde(default)]
    pub artifact_uri: String,
    #[serde(default)]
    pub entrypoint: String,
    #[serde(default)]
    pub requirements_uri: String,
    #[serde(default)]
    pub metadata: Value,
}

impl ModelAdapterDescriptor {
    pub fn has_signal(&self) -> bool {
        !self.kind.trim().is_empty()
            || !self.framework.trim().is_empty()
            || !self.flavor.trim().is_empty()
            || !self.runtime.trim().is_empty()
            || !self.loader.trim().is_empty()
            || !self.artifact_uri.trim().is_empty()
            || !self.entrypoint.trim().is_empty()
            || !self.requirements_uri.trim().is_empty()
            || !self.metadata.is_null()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct RegistrySourceDescriptor {
    #[serde(default)]
    pub system: String,
    #[serde(default)]
    pub model_name: String,
    #[serde(default)]
    pub model_version: String,
    #[serde(default)]
    pub stage: String,
    #[serde(default)]
    pub uri: String,
    #[serde(default)]
    pub metadata: Value,
}

impl RegistrySourceDescriptor {
    pub fn has_signal(&self) -> bool {
        !self.system.trim().is_empty()
            || !self.model_name.trim().is_empty()
            || !self.model_version.trim().is_empty()
            || !self.stage.trim().is_empty()
            || !self.uri.trim().is_empty()
            || !self.metadata.is_null()
    }
}

File: openfoundry-go/libs/ml-kernel-go/models/interop.go
package models

import (
	"encoding/json"
	"strings"
)

// ExternalTrackingSource captures references to an external ML
// tracking system (MLflow, Weights & Biases, …).
type ExternalTrackingSource struct {
	System                  string              `json:"system,omitempty"`
	Project                 string              `json:"project,omitempty"`
	ExperimentName          string              `json:"experiment_name,omitempty"`
	RunID                   string              `json:"run_id,omitempty"`
	RunName                 string              `json:"run_name,omitempty"`
	RunURI                  string              `json:"run_uri,omitempty"`
	ArtifactURI             string              `json:"artifact_uri,omitempty"`
	ModelURI                string              `json:"model_uri,omitempty"`
	RegisteredModelName     string              `json:"registered_model_name,omitempty"`
	RegisteredModelVersion  string              `json:"registered_model_version,omitempty"`
	Framework               string              `json:"framework,omitempty"`
	Flavor                  string              `json:"flavor,omitempty"`
	Stage                   string              `json:"stage,omitempty"`
	Tags                    json.RawMessage     `json:"tags,omitempty"`
	Params                  json.RawMessage     `json:"params,omitempty"`
	Metrics                 []MetricValue       `json:"metrics,omitempty"`
	Artifacts               []ArtifactReference `json:"artifacts,omitempty"`
	Metadata                json.RawMessage     `json:"metadata,omitempty"`
}

// HasSignal returns true when at least one field carries data.
// Mirrors the Rust impl ExternalTrackingSource::has_signal.
func (s *ExternalTrackingSource) HasSignal() bool {
	return strings.TrimSpace(s.System) != "" ||
		strings.TrimSpace(s.Project) != "" ||
		strings.TrimSpace(s.ExperimentName) != "" ||
		strings.TrimSpace(s.RunID) != "" ||
		strings.TrimSpace(s.RunName) != "" ||
		strings.TrimSpace(s.RunURI) != "" ||
		strings.TrimSpace(s.ArtifactURI) != "" ||
		strings.TrimSpace(s.ModelURI) != "" ||
		strings.TrimSpace(s.RegisteredModelName) != "" ||
		strings.TrimSpace(s.RegisteredModelVersion) != "" ||
		strings.TrimSpace(s.Framework) != "" ||
		strings.TrimSpace(s.Flavor) != "" ||
		strings.TrimSpace(s.Stage) != "" ||
		len(s.Metrics) > 0 ||
		len(s.Artifacts) > 0 ||
		!isJSONNull(s.Params) ||
		!isJSONNull(s.Tags) ||
		!isJSONNull(s.Metadata)
}

// ModelAdapterDescriptor describes a custom model adapter (loader).
type ModelAdapterDescriptor struct {
	Kind            string          `json:"kind,omitempty"`
	Framework       string          `json:"framework,omitempty"`
	Flavor          string          `json:"flavor,omitempty"`
	Runtime         string          `json:"runtime,omitempty"`
	Loader          string          `json:"loader,omitempty"`
	ArtifactURI     string          `json:"artifact_uri,omitempty"`
	Entrypoint      string          `json:"entrypoint,omitempty"`
	RequirementsURI string          `json:"requirements_uri,omitempty"`
	Metadata        json.RawMessage `json:"metadata,omitempty"`
}

// HasSignal mirrors the Rust impl.
func (d *ModelAdapterDescriptor) HasSignal() bool {
	return strings.TrimSpace(d.Kind) != "" ||
		strings.TrimSpace(d.Framework) != "" ||
		strings.TrimSpace(d.Flavor) != "" ||
		strings.TrimSpace(d.Runtime) != "" ||
		strings.TrimSpace(d.Loader) != "" ||
		strings.TrimSpace(d.ArtifactURI) != "" ||
		strings.TrimSpace(d.Entrypoint) != "" ||
		strings.TrimSpace(d.RequirementsURI) != "" ||
		!isJSONNull(d.Metadata)
}

// RegistrySourceDescriptor points at a registry-system reference.
type RegistrySourceDescriptor struct {
	System       string          `json:"system,omitempty"`
	ModelName    string          `json:"model_name,omitempty"`
	ModelVersion string          `json:"model_version,omitempty"`
	Stage        string          `json:"stage,omitempty"`
	URI          string          `json:"uri,omitempty"`
	Metadata     json.RawMessage `json:"metadata,omitempty"`
}

// HasSignal mirrors the Rust impl.
func (r *RegistrySourceDescriptor) HasSignal() bool {
	return strings.TrimSpace(r.System) != "" ||
		strings.TrimSpace(r.ModelName) != "" ||
		strings.TrimSpace(r.ModelVersion) != "" ||
		strings.TrimSpace(r.Stage) != "" ||
		strings.TrimSpace(r.URI) != "" ||
		!isJSONNull(r.Metadata)
}

// isJSONNull treats nil/empty/literal "null" raw messages as null —
// mirrors Rust serde_json::Value::is_null() for the omitempty inputs.
func isJSONNull(b json.RawMessage) bool {
	if len(b) == 0 {
		return true
	}
	s := strings.TrimSpace(string(b))
	return s == "" || s == "null"
}
